README.md
# RAET (Reliable Asynchronous Event Transport) Protocol

## Motivation

Modern large scale distributed application architectures, wherein components are
distributed across the internet on multiple hosts and multiple CPU cores, are often
based on a messaging or event bus that allows the various distributed components
to communicate asynchronously with each other. Typically the messaging bus is
some form of messaging queue service such as AMQP or ZeroMQ. The message bus supports
what is commonly referred to as a publish/subscribe methodology for information
exchange.

While there are many advantages to a full featured message queuing service,
one of the disadvantages is the inability to manage performance at scale.

A message queuing service performs two distinct but complementary functions.

- The first is asynchronous transport of messages over the internet.
- The second is message queue management, that is, the identification, tracking,
storage, and distribution of messages between publishers and subscribers via queues.

One of the advantages of a message queuing service for many applications is that
the service hides behind an API, the complexities of queue management from the clients.
The disadvantage is that at scale, where the volume of messages, the
timing of messages, and the associated demands on memory, network, and cpu capacity
become critical, the client has little ability to tune the service for performance.
Often MQ services become bottlenecks for the distributed application.
The more complicated MQ services, like AMQP, tend to be unreliable under load.

Separating the function of network transport of asynchrounous event from the
function of message queue management allows independant tuning at scale of each function.

Most if not all of the MQ services are based on TCP/IP for transport.
TCP/IP adds significant latency to the network communications and is therefore
not well suited for the asynchronous nature of distibuted event driven application
communications. This is primarily due to the way TCP/IP handles connection setup
and teardown as well as failed connections in order to support streams. Fundamentally
TCP/IP is optomized for sending large contiguous data streams not many small
aynchronous events or messages. While not a problem for small scale systems,
the differences in the associated traffic characteristics can become problematic
at scale.

Because UDP/IP has lower latency and is connectionless, it is much better suited
to many small asynchronous messages and scales better. The drawback of bare UDP/IP
is that it is not reliable. What is needed, therefore, is a tuned transport
protocol that adds reliability to UDP/IP without sacrificing latency and scalability.
A transactioned protocol, is much more appropriate for providing reliablity to
asynchronous event transport than a streaming protocol.

Moreover, because most MQ services are based on TCP/IP they tend to also use
HTTP and therefore TLS/SSL for secure communications. While using HTTP provides
easy integration with web based systems, it can become problematic for high performant systems
Furthermore, TLS is also problematic as a security system both from performance
and vulnerabilty aspects.

Elliptic Curve Cryptography, on the other hand, provides increases in security
with lower performance requirements relative to over other approaches.
LibSodium provides an open source Elliptic Curve Cryptographic library with support
for both authentication and encryption. The CurveCP protocol is based on LibSodium
and provides a handshake protocol for bootstrapping secure network exchanges of information.

Finally, one of the best ways to manage and fine tune processor resources
(cpu, memory, network) in distributed concurrent event driven applications is to use
something called micro-threads. A microthread is typically an in-language feature
that allows logical concurrency with no more overhead than a function call.
Micro threading uses cooperative multi-tasking instead of threads and/or processes
and avoids many of the complexities of resource contention, context switching,
and interprocess communications while providing much higher total performance.

Because all the cooperative micro-threads run in one process, a simple micro-threaded
application is limited to one CPU core. To enable full utilization of all CPU
cores, the application needs to be able to run at least one process per CPU core.
This requires same host inter-process communications. But unlike the conventional
approach to multi-processing  where there is of one process per logical concurrent
function, a micro-threaded multi-process application has instead one micro-thread
per logical concurrent function and the total number of micro-threads
is distributed amoungst a minimal number of processes, no more than the number of
cpu cores. This optimizes the use of the cpu power while minimizes the overhead of
process context switching.

An example of a framework that uses this type of micro-threaded but multi-process
architecture is Erlang. Indeed, the success of the Erlang model provided
support for the viability of the RAET approach.
Indeed, one might ask, why not use Erlang? Unfortunately, the Erlang ecosystem is
somewhat limited in comparison to Python's and the language itself uses what one
might describe as a very unfortunate syntax.
One of the design objectives behine RAET was to leverage existing Python expertise
and the richness of the Python ecosystem but still be able to develop distributed
applications using a micro-threaded multi-process architectural model. The goal was
to combine the best of both worlds.

RAET is designed to provide secure reliable scalable asynchronous message/event
transport over the internet in a micro-threaded multi-process application framework
that uses UDP for interhost communication and LibSodium for authentication, encryption
and the CurveCP handshake for secure bootstrap.

The queue management and micro-threaded application support is provided by Ioflo.
RAET is a complementary project to Ioflo in that RAET enables multiple Ioflo
applications to work together over a network as part of a distributed application.

The primary use case and motivating problem that resulted in the development of RAET
was the need to enable SaltStack to scale better. SaltStack is a remote execution
and configuration management platform written in Python. SaltStack uses ZeroMQ (0MQ)
as its message bus or message queuing service. ZeroMQ is based on TCP/IP so suffers from
the aforementioned latency and non-asynchronicity issues of TCP/IP based architectures.
Moreover because ZeroMQ integrates queue management and transport in a monolithic way
with special "sockets", tuning the performance of the queuing independent of the transport
at scale becomes problematic. Tracing down bugs can also be problematic.


## Installation

Current raet is provided as a PyPi package. To install on a unix based system
use pip.

``` bash

$ pip install raet


```

on OS X

``` bash
$ sudo pip install raet

```



## Introduction

Currently RAET supports two types of communication.

- Host to host communication over UDP/IP sockets
- Same host interprocess communication over Unix Domain (UXD) Sockets

The architecture of a RAET based application is shown in the figure below:

![Diagram 1](docs/images/RaetMetaphor.png?raw=true)

##Naming Metaphor for Components

The following naming metaphor is designed to consistent but not conflicting with Ioflo

### Road, Estates, Main Estate

- The UDP channel is  a “Road"
- The members of a Road are “Estates”  (as in real estate lots that front the road)
- Each Estate has a unique UDP Host Port address “ha” , a unique  string “name”
    and unique numerical ID “eid".
- One Estate on the Road is the “Main” Estate
- The Main Estate is responsible for allowing other estates to join the Road
    via the Join (Key Exchange)  and Allow (CurveCP) transactions
- The Main Estate is also responsible for routing messages between other Estates

### Lane, Yards, Main Yard

- Within each Estate may be a “Lane”.  This is a UXD channel.
- The members of a Lane are “Yards”  (as in subdivision plots within the Estate)
- Each Yard on a Lane has a unique UXD file name host address “ha’, and a unique string “name”.
    There is also a numerical Yard ID “yid" that the class uses to generate yard
    names but it is not an attribute of the Yard instance.
- The Lane name is also used with the Yard Name to form a unique Filename that
    is the ha of the UXD
- One Yard on a Lane is  the “Main” Yard
- The Main Yard is responsible for forming the Lane and permitting other Yards
    to be on the lane. There is yet no formal process for this.
    Currently there is a flag that will drop packets from any Yard that is not
    already in the list of Yards maintained by the Main yard.
    Also file permissions can be used to prevent spurious Yards from communicating
    with the Main Yard.
- The Main Yard is responsible for routing messages between other yards on the Lane


## IoFlo Execution

- Each Estate UDP interface is run via a RoadStack (UDP sockets)
    which is run within the context of an IoFlo House
    (so think of the House that runs the UDP Stack as the Manor House of the Estate)
- Each Yard UXD interface is run via a LaneStack (Unix domain socket)
    which is run within the context of an IoFlo House
    (so think of Houses that run UXD stacks as accessory Houses (Tents, Shacks) on the Estate)
- The "Manor" House is special in that it runs both the UDP stack for the Estate
     and also runs the UXD Stack for the Main Yard
- The House that runs the Main Estate UDP Stack can be thought of as Mayor’s House
- Within the context of a House is a Data Store. Shares in the Store are Addressed
    by the unique Share Name which is a dotted path

## Routing

Given the Ioflo execution architecture described above, routing is performed as follows:

- In order to address a specific Estate, the Estate Name is required
- In order to address a specific Yard within an Estate, the Yard Name is required
- In order to address a specific Queue within a House, the Share Name is required

The UDP stack maps Estate Name to UDP HA and Estate ID
The UXD stack maps Yard Name to UXD HA
The Store of any IoFlo behavior maps Share Name to Share reference

Therefore Routing
from: a source identified by
    a queue in a source Share,
    in a source Yard,
    in a source Estate
to: a destination identified by
    a queue, in a destination Share,
    in a destination Yard,
    in a destination Estate

requires two Triples, one for Source and one for Destination

Source
(Estate Name, Yard Name, Share Name)

Destination
(Estate Name, Yard Name, Share Name)

If any element of the Triple is None or Empty then a Default is used.

Below is an example of a Message Body that has the Routing information it it.


```python
    estate = 'minion1'
    stack0 = stacking.StackUxd(name='lord', lanename='cherry', yid=0)
    stack1 = stacking.StackUxd(name='serf', lanename='cherry', yid=1)
    yard = yarding.Yard( name=stack0.yard.name, prefix='cherry')
    stack1.addRemoteYard(yard)

    src = (estate, stack1.yard.name, None)
    dst = (estate, stack0.yard.name, None)
    route = odict(src=src, dst=dst)
    msg = odict(route=route, stuff="Serf to my lord. Feed me!")
    stack1.transmit(msg=msg)

    timer = Timer(duration=0.5)
    timer.restart()
    while not timer.expired:
        stack0.serviceAll()
        stack1.serviceAll()


    lord Received Message
    {
        'route':
        {
            'src': ['minion1', 'yard1', None],
            'dst': ['minion1', 'yard0', None]
        },
        'stuff': 'Serf to my lord. Feed me!'
    }
````



## Details of UDP/IP Raet Protocol

The UDP Raet protocol is based on a coding metaphor naming convention, that is, of
estates attached to a road. The core objects are provided in the following package:
raet.road

### Road Raet Production UDP/IP Ports

Manor Estate 4505
Other Estates 4510

### Packet Data Format

The data used to initialize a packet is an ordered dict with several fields
most of the fields are shared with the header data format below so only the
unique fields are shown here.

### Unique Packet data fields

    sh: source host ip address (ipv4)
    sp: source ip port
    dh: destination host ip address (ipv4)
    dp: destination host ip port

### Header Data Format.
The .data in the packet header is an ordered dict  which is used to either
create a packet to transmit
or holds the field from a received packet.
What fields are included in a header is dependent on the header kind.

### Header encoding

There are three header encoding formats currently supported.

- RAET Native. This is an minimized ascii test format that optimizes the tradeoff between
easy readability and size. This is the default.

- JSON. This is the most verbose format but has the advantage of compatibility.

- Binary. This is not yet implemented. Once the protocol reaches a more mature state
and its not likely that there will be any header changes (or very infrequent) then
a binary format that minimizes size will be provided.

When the head kind is json = 0, then certain optimizations are used to minimize
the header length.
- The header field keys are two bytes long
- If a header field value is the default then the field is not included
- Lengths are encoded as hex strings
- The flags are encoded as a double char hex string in field 'fg'


### Header Data Fields

    ri: raet id Default 'RAET'
    vn: Version (Version) Default 0
    pk: Packet Kind (PcktKind)
    pl: Packet Length (PcktLen)
    hk: Header kind   (HeadKind) Default 0
    hl: Header length (HeadLen) Default 0

    se: Source Estate ID (SEID)
    de: Destination Estate ID (DEID)
    cf: Correspondent Flag (CrdtFlag) Default 0
    bf: BroadCast Flag (BcstFlag)  Default 0

    si: Session ID (SID) Default 0
    ti: Transaction ID (TID) Default 0
    tk: Transaction Kind (TrnsKind)

    dt: Datetime Stamp  (Datetime) Default 0
    oi: Order index (OrdrIndx)   Default 0

    wf: Waiting Ack Flag    (WaitFlag) Default 0
        Next segment or ordered packet is waiting for ack to this packet
    ml: Message Length (MsgLen)  Default 0
        Length of message only (unsegmented)
    sn: Segment Number (SgmtNum) Default 0
    sc: Segment Count  (SgmtCnt) Default 1
    sf: Segment Flag  (SgmtFlag) Default 0
        This packet is part of a segmented message
    af: All Flag (AllFlag) Default 0
        Resend all segments not just one

    bk: Body kind   (BodyKind) Default 0
    ck: Coat kind   (CoatKind) Default 0
    fk: Footer kind   (FootKind) Default 0
    fl: Footer length (FootLen) Default 0

    fg: flags  packed (Flags) Default '00' hs
         2 char Hex string with bits (0, 0, af, sf, 0, wf, bf, cf)
         Zeros are TBD flags


### Body Data Format

The Body .data is a Mapping that is serialized using either JSON or MSGPACK

### Packet Parts

Each packet has 4 parts some of which may be empty. These are:
- Head
- Body
- Coat
- Tail

The Head is manditory and provides the header fields that are needed to process the
packet.

The Tail provides the authentication signature that is used to verify the source of
the packet and that its contents have not been tampered with.

The Body is the contents of the packet. Some packets such as Acks and Nacks don't
need a body. The Body is a serialized Python dictionary typically and ordered dictionary
so that parsing and debugging has a consistent view of the ordering of the fields
in the body.

The Coat is the encrypted version of the body. The encryption type is CurveCP based.
If the Coat is provided then the Body is encapsulated within the Coat Part.



### Header Details

#### JSON Encoding

Header is the ASCII Safe JSON encoding of a Python ordered dictionary.
Header termination is an empty line given by double pair of carriage-return linefeed
characters.

/r/n/r/n
10 13 10 13
ADAD
1010 1101 1010 1101

Carriage-return and newline characters cannot appear in a JSON encoded
string unless they are escaped with backslash, so the 4 byte combination is illegal
in valid JSON that does not have multi-byte unicode characters so it makes it a
uniquely identifiable header termination.

These means the header must be ascii safe  so no multibyte utf-8 strings are
allowed in the header.

#### Native Encoding

The header consists of newline delimited lines. Each header line consists
of a two character field identifier followed by a space followed by the value
of the field as ascii hex encoded binary followed by newline.
The Header end is indicated by a blank line, that is, a double newline character.
Example


ChangeLog.md
------------------
CHANGE LOG
-------------------


-----------
201707XX
----------
v0.6.9

Added MAX_MISSEDS_RESEND = min((UDP_MAX_PACKET_SIZE - MAX_HEAD_SIZE - 110) // 6, 64
for max number of misseds in Messengent.resend packet instead of just 64

Add codesponsor


-----------
20170721
----------
v0.6.8

Added .veritive flag to RoadStack to enable better control of accepting signed
packets

-----------
20170313
----------
v0.6.7

Fixed tabs to spaces
Added newRemote wrapper method aroung RemoteEstate() to allow for custom validation
  of arguments  when creating a new remote
Fixed some test race timing conditions on faster computer with python3.6
Fixed typo in stacking.py bug


-----------
20161228
----------
v0.6.6

Support for ipv6
Bug fixes
More detail in console log messages associated with stale sessions

-----------
20151018
----------
v0.6.5

Fixed to work with python 3.5 withoug simplejson. Removed encoding parameter.
Fixed up unittests
Add local test runners for sub section tests to run part of unit tests
Added requirement of simpljson for python2.6 for object pairs hook



-----------
20151017
----------
v0.6.4

Fixed issue with windows errno
Minor corrections
Fixed imports of ioflo to be compat with ioflo 1.2.3
Removed unused imports and normalized some imports


-----------
20150310
----------
v0.6.3

Changed imports ioflo.base.nonblocking to be compat with ioflo 1.2.1

-----------
20150212
----------
v0.6.1

Updated tutorial
fixed log message


-----------
20150212
----------
v0.6.0

Added enum support using python 3.4 enums via backport enum34
raeting.py NamedTuples replaced with enums

-----------
20150211
----------
v0.5.3

Added tutorial documents
Fixed merge error
Fixed bugs


-----------
20150204
----------
v0.5.2

Updated setup.py dependencies

-----------
20150204
----------
v0.5.1

Windows impossible file check
More coverage tests

-----------
20150204
----------
v0.5.0

Python3 support



-----------
20150201
----------
v0.4.7

Added support for burst size limits in the Message transaction to deal with
limits on UDP receive buffers on some hosts. Significant changes to how the
Messenger and Messagent transactions work.

Added more units tests for messaging.


-----------
20150128
----------
v0.4.6

Silently handles errno.ECONNRESET on transmit in Stack._handleOneTx for windows



-----------
20150128
----------
v0.4.5

Support for simultaneous vacuous joins between same two estates
Correctly terminates one of the joins and allows other to complete
Bugfixes
More tests

-----------
20150120
----------
v0.4.4

Compatibility with ioflo 1.1.5

-----------
20150120
----------
v0.4.3

Minor bug fixes
More units tests

Changed semantics for initial alived state. As soon as estate completes
initial bootstrap to allowed state it is also alived instead of waiting for
manage alive transactions to complete. This enables messaging to commence
immediately upon allowed.


-----------
20141201
----------
v0.4.2

fixed windows compat for autoaccept yard



---------
20141125
----------
v0.4.0

Changed the allow transaction to bigendian structs not little endian
This will break interoperability with older versions of raet.
Msgpack is already bigendian and future binary packet header for raet
should be big endian.


more coverage unit tests
fixed some bugs
updated log messages
updated comments
updated indenting
added more presence unit tests

---------
20141029
----------
v0.3.8

Added more unit tests
Updated and fixed some tests
Joiner and Joinent transactions now persist local data (puid) upon successful completion


---------
20141024
----------
v0.3.7

Added lots of unit tests for join (Joiner and Joinent) transaction
Fixed some bugs
Added better support (but not complete ) for windows



---------
20141009
----------
v0.3.6

Made default timeout for Messenger and Messengent transaction to be 0.0, that is,
never

Added timout parameter to RoadStack.transmit() so timeout can be set on a
message transaction by transaction basis
Added timeout to txMsgs deque triple for RoadStack that is, (msg, uid, timeout)



---------
20141008
----------
v0.3.5

Fixed bug in RoadKeep.loadRemoteRoleData()


---------
20141006
----------
v0.3.4

Added socket error handling for additional failure modes in Stack._handleOneTx
for sending

---------
20141006
----------
v0.3.3

Fixed Python3 incompatabilities



---------
20141002
----------
v0.3.02

Fixed bug in allow transaction that prevented allow on restart
Fixed issue with unicode fqdn on debian


---------
20140924
----------
v0.3.01

New support for peer RAET channel. Changed semantics of how Join transactions
works. This will allows peer to peer joining

---------
20140729
---------

v0.2.11

Lookup remote in stack handleOneRx by using packet.index since it already
substitutes ha if se is zero. Then also lookup remote in haRemotes if not found
in .uidRemotes

Update joinent transaction

Change the way persistence reaping is handled. Instead of removing remote from
memory, mark it with .reaped attribute. This avoide problematic use case of
having to restoreRemote from disk when doing lookup in Joinent.
Update .manage

Also unreap anytime a packet is received from a remote

---------
20140724
---------

v0.2.10

Transactions now referenced and processed per remote not at stack.transactions
sets stage for later doing broadcast transactions

---------
20140724
---------

v0.2.09

Finished support for role keep files in Road

---------
20140723
---------

v0.2.08

Support for msgpack in Keep files
more support for role in Road estates
fixes in join transaction

---------
20140722
---------

v0.2.07

bugfix

---------
20140722
---------

v0.2.06

If verify keep fails then clear the keep file. This autoclears cache
when upgrading format of keep file.
Also more support for role.

---------
20140721
---------

v0.2.05

Refactored odicts indexing access attributes for remotes so can now access
by uid, by name, or by ha. Updated associated accessors

raet/raeting.py
# -*- coding: utf-8 -*-
'''
raeting module provides constants and values for the RAET protocol

Production Ports for Raet
Master 4505
Minion(s) 4510

Packet Data Format.
The data used to initialize a packet is an ordered dict with several fields
most of the fields are shared with the header data format below so only the
unique fields are shown here.

Unique Packet data fields

    sh: source host ip address (ipv4) Default: ''
    sp: source ip port                Default: 7532
    dh: destination host ip address (ipv4) Default: '127.0.0.1'
    dp: destination host ip port           Default 7532

Header Data Format.
The .data in the packet header is an ordered dict  which is used to either
create a packet to transmit
or holds the field from a received packet.
What fields are included in a header is dependent on the header kind.

Header encoding.
    When the head kind is json = 0, then certain optimizations are
    used to minimize the header length.
        The header field keys are two bytes long
        If a header field value is the default then the field is not included
        Lengths are encoded as hex strings
        The flags are encoded as a double char hex string in field 'fg'

header data =
{
    ri: raet id Default 'RAET'
    vn: Version (Version) Default 0
    pk: Packet Kind (PcktKind)
    pl: Packet Length (PcktLen)
    hk: Header kind   (HeadKind) Default 0
    hl: Header length (HeadLen) Default 0

    se: Source Estate ID (SEID)
    de: Destination Estate ID (DEID)
    cf: Correspondent Flag (CrdtFlag) Default 0
    bf: BroadCast Flag (BcstFlag)  Default 0
    nf: NAT Flag (NatFlag) Default 0
    df: Dynamic IP Flag (DynFlag) Default 0
    vf: IPv6 Flag (IP) (Ipv6Flag) Default 0

    si: Session ID (SID) Default 0
    ti: Transaction ID (TID) Default 0
    tk: Transaction Kind (TrnsKind)

    dt: Datetime Stamp  (Datetime) Default 0
    oi: Order index (OrdrIndx)   Default 0

    wf: Waiting Ack Flag    (WaitFlag) Default 0
        Next segment or ordered packet is waiting for ack to this packet
    ml: Message Length (MsgLen)  Default 0
        Length of message only (unsegmented)
    sn: Segment Number (SgmtNum) Default 0
    sc: Segment Count  (SgmtCnt) Default 1
    sf: Segment Flag  (SgmtFlag) Default 0
        This packet is part of a segmented message
    af: Again Flag (AgnFlag) Default 0
        This segment is being sent again

    bk: Body kind   (BodyKind) Default 0
    ck: Coat kind   (CoatKind) Default 0
    fk: Footer kind   (FootKind) Default 0
    fl: Footer length (FootLen) Default 0

    fg: flags  packed (Flags) Default '00' hs
         2 char Hex string with bits (vf, df, nf, af, sf, wf, bf, cf)
         Zeros are TBD flags
}

Body Data Format
The Body .data is a Mapping

Body Encoding
    When the body kind is json = 0, then the .data is json encoded

Body Decoding


'''

# pylint: disable=C0103

# Import python libs
import struct
import enum

# Import ioflo libs
from ioflo.aid.odicting import odict

# Import raet libs
# pylint: disable=wildcard-import,unused-wildcard-import,redefined-builtin
from .abiding import *  # import globals
# pylint: enable=wildcard-import,unused-wildcard-import,redefined-builtin

# Used to comput session id wrap around where valid sid is >= modulo N given by
# (((new - old) % 0x100000000) < (0x100000000 // 2))
# N//2 = 0x80000000
SID_WRAP_MODULO = 0x100000000  # session id wraps modulo N = 2^32 = 0x100000000
SID_WRAP_DELTA = 0x80000000    # session id >= delta at N//2 = 0x80000000
SID_ROLLOVER = 0xffffffff      # session id rolls over at modulo (N-1) -= 2^32 -1 = 0xffffffff

RAET_PORT = 7530
RAET_TEST_PORT = 7531
DEFAULT_SRC_HOST = ''
DEFAULT_DST_HOST = '127.0.0.1'

UDP_MAX_DATAGRAM_SIZE = (2 ** 16) - 1  # 65535
UDP_MAX_SAFE_PAYLOAD = 548  # IPV4 MTU 576 - udp headers 28
# IPV6 MTU is 1280 but headers are bigger
UDP_MAX_PACKET_SIZE = min(1024, UDP_MAX_DATAGRAM_SIZE)  # assumes IPV6 capable equipment
UXD_MAX_PACKET_SIZE = (2 ** 16) - 1  # 65535
MAX_SEGMENT_COUNT = (2 ** 16) - 1  # 65535
MAX_MESSAGE_SIZE = min(67107840, UDP_MAX_PACKET_SIZE * MAX_SEGMENT_COUNT)
MAX_HEAD_SIZE = 255
MAX_MISSEDS_RESEND = min((UDP_MAX_PACKET_SIZE - MAX_HEAD_SIZE - 110) // 6, 64)

JSON_END = b'\r\n\r\n'
HEAD_END = b'\n\n'

VERSIONS = odict([('0.1', 0)])
VERSION_NAMES = odict((v, k) for k, v in VERSIONS.iteritems())
VERSION = VERSIONS.values()[0]

HELLO_PACKER = struct.Struct('!64s32s80s24s')  # curvecp allow trans bodies
COOKIESTUFF_PACKER = struct.Struct('!32sLL24s')
COOKIE_PACKER = struct.Struct('!80s24s')
INITIATESTUFF_PACKER = struct.Struct('!32s48s24s128s')
INITIATE_PACKER = struct.Struct('!32s24s248s24s')


def get_exception_error(ex):
    '''
    Return the error code from an exception
    '''
    if hasattr(ex, 'errno'):
        return ex.errno
    elif hasattr(ex, 'winerror'):
        return ex.winerror
    else:
        emsg = "Cannot find error code in exception: {0}".format(ex)
        raise TypeError(emsg)


@enum.unique
class HeadKind(enum.IntEnum):
    '''
    Integer Enums of Head Kinds
    '''
    raet = 0
    json = 1
    binary = 2
    unknown = 255


@enum.unique
class BodyKind(enum.IntEnum):
    '''
    Integer Enums of Body Kinds
    '''
    nada = 0
    json = 1
    raw = 2
    msgpack = 3
    unknown = 255


@enum.unique
class FootKind(enum.IntEnum):
    '''
    Integer Enums of Foot Kinds
    '''
    nada = 0
    nacl = 1
    sha2 = 2
    crc64 = 3
    unknown = 255


class FootSize(enum.IntEnum):
    '''
    Integer Enums of Foot Sizes in bytes
    '''
    nada = 0
    nacl = 64
    sha2 = 256
    crc64 = 8
    unknown = 0


@enum.unique
class CoatKind(enum.IntEnum):
    '''
    Integer Enums of Coat Kinds
    '''
    nada = 0
    nacl = 1
    crc16 = 2
    crc64 = 3
    unknown = 255


class TailSize(enum.IntEnum):
    '''
    Integer Enums of Tail Sizes in bytes
    '''
    nada = 0
    nacl = 24
    crc16 = 2
    crc64 = 8
    unknown = 0


@enum.unique
class TrnsKind(enum.IntEnum):
    '''
    Integer Enums of Transaction Kinds
    '''
    message = 0
    join = 1
    bind = 2
    allow = 3
    alive = 4
    unknown = 255


@enum.unique
class PcktKind(enum.IntEnum):
    '''
    Integer Enums of Packet Kinds
    '''
    message = 0
    ack = 1
    nack = 2
    resend = 3
    request = 4
    response = 5
    hello = 6
    cookie = 7
    initiate = 8
    unjoined = 9
    unallowed = 10
    renew = 11
    refuse = 12
    reject = 13
    pend = 14
    done = 15
    unknown = 255


@enum.unique
class Acceptance(enum.IntEnum):
    '''
    Integer Enums of Acceptances
    '''
    pending = 0
    accepted = 1
    rejected = 2


@enum.unique
class AutoMode(enum.IntEnum):
    '''
    Integer Enums of Auto Modes
    '''
    never = 0
    once = 1
    always = 2


@enum.unique
class PackKind(enum.IntEnum):
    '''
    Integer Enums of Pack Kinds for Lane Pages
    '''
    json = 0
    pack = 1


# head fields that may be included in packet header if not default value
PACKET_DEFAULTS = odict([
                            ('sh', DEFAULT_SRC_HOST),
                            ('sp', RAET_PORT),
                            ('dh', DEFAULT_DST_HOST),
                            ('dp', RAET_PORT),
                            ('ri', 'RAET'),
                            ('vn', 0),
                            ('pk', 0),
                            ('pl', 0),
                            ('hk', 0),
                            ('hl', 0),
                            ('se', 0),
                            ('de', 0),
                            ('cf', False),
                            ('bf', False),
                            ('nf', False),
                            ('df', False),
                            ('vf', False),
                            ('si', 0),
                            ('ti', 0),
                            ('tk', 0),
                            ('dt', 0),
                            ('oi', 0),
                            ('wf', False),
                            ('sn', 0),
                            ('sc', 1),
                            ('ml', 0),
                            ('sf', False),
                            ('af', False),
                            ('bk', 0),
                            ('ck', 0),
                            ('fk', 0),
                            ('fl', 0),
                            ('fg', '00'),
                      ])

PACKET_FIELDS = ['sh', 'sp', 'dh', 'dp',
                 'ri', 'vn', 'pk', 'pl', 'hk', 'hl',
                 'se', 'de', 'cf', 'bf', 'nf', 'df', 'vf', 'si', 'ti', 'tk',
                 'dt', 'oi', 'wf', 'sn', 'sc', 'ml', 'sf', 'af',
                 'bk', 'ck', 'fk', 'fl', 'fg']

PACKET_HEAD_FIELDS = ['ri', 'vn', 'pk', 'pl', 'hk', 'hl',
               'se', 'de', 'cf', 'bf', 'nf', 'df', 'vf', 'si', 'ti', 'tk',
               'dt', 'oi', 'wf', 'sn', 'sc', 'ml', 'sf', 'af',
               'bk', 'bl', 'ck', 'cl', 'fk', 'fl', 'fg']

PACKET_FLAGS = ['vf', 'df', 'nf', 'af', 'sf', 'wf', 'bf', 'cf']
PACKET_FLAG_FIELDS = ['vf', 'df', 'nf', 'af', 'sf', 'wf', 'bf', 'cf']

PACKET_FIELD_FORMATS = odict([
                    ('ri', '.4s'),
                    ('vn', 'x'),
                    ('pk', 'x'),
                    ('pl', '04x'),
                    ('hk', 'x'),
                    ('hl', '02x'),
                    ('se', 'x'),
                    ('de', 'x'),
                    ('cf', ''),
                    ('bf', ''),
                    ('nf', ''),
                    ('df', ''),
                    ('vf', ''),
                    ('si', 'x'),
                    ('ti', 'x'),
                    ('tk', 'x'),
                    ('dt', 'f'),
                    ('oi', 'x'),
                    ('wf', ''),
                    ('sn', 'x'),
                    ('sc', 'x'),
                    ('ml', 'x'),
                    ('sf', ''),
                    ('af', ''),
                    ('bk', 'x'),
                    ('ck', 'x'),
                    ('fk', 'x'),
                    ('fl', 'x'),
                    ('fg', '.2s'),
              ])

# head fields that may be included in page header if not default value
PAGE_DEFAULTS = odict([
                        ('ri', 'RAET'),
                        ('vn', 0),
                        ('pk', 0),
                        ('sn', ''),
                        ('dn', ''),
                        ('si', '000000000000000000'),
                        ('bi', 0),
                        ('pn', 0),
                        ('pc', 1),
                      ])

PAGE_FIELD_FORMATS = odict([
                            ('ri', '.4s'),
                            ('vn', 'x'),
                            ('pk', 'x'),
                            ('sn', 's'),
                            ('dn', 's'),
                            ('si', '.18s'),
                            ('bi', 'x'),
                            ('pn', '04x'),
                            ('pc', '04x'),
                           ])

PAGE_FIELDS = ['ri', 'vn', 'pk', 'sn', 'dn', 'si', 'bi', 'pn', 'pc']


class RaetError(Exception):
    '''
    Exceptions in RAET Protocol processing

       usage:
           emsg = "Invalid unique id '{0}'".format(uid)
           raise raeting.RaetError(emsg)
    '''
    def __init__(self, message=None):
        self.message = message  # description of error
        super(RaetError, self).__init__(message)

    def __str__(self):
        return "{0}: {1}.\n".format(self.__class__.__name__, self.message)


class StackError(RaetError):
    '''
       Exceptions in RAET stack processing

       Usage:
            emsg = "Invalid unique id '{0}'".format(uid)
            raise raeting.StackError(emsg)
    '''
    pass


class EstateError(RaetError):
    '''
       Exceptions in RAET estate processing

       Usage:
            emsg = "Invalid unique id '{0}'".format(uid)
            raise raeting.EstateError(emsg)
    '''
    pass


class TransactionError(RaetError):
    '''
       Exceptions in RAET transaction processing

       Usage:
            emsg = "Invalid uniqu id '{0}'".format(uid)
            raise raeting.TransactionError(emsg)
    '''
    pass


class PacketError(RaetError):
    '''
       Exceptions in RAET packet processing

       Usage:
            emsg = "Invalid unique id '{0}'".format(uid)
            raise raeting.PacketError(emsg)
    '''
    pass


class PacketSizeError(PacketError):
    '''
       Packet too large error needs to be segmented

       Usage:
            emsg = "Packet size {0} too large needs to be segmented".format(size)
            raise raeting.PacketSizeError(emsg)
    '''
    pass


class KeepError(RaetError):
    '''
       Exceptions in RAET keep processing

       Usage:
            emsg = "Invalid unique id '{0}'".format(uid)
            raise raeting.KeepError(emsg)
    '''
    pass


class YardError(RaetError):
    '''
       Exceptions in RAET yard processing

       Usage:
            emsg = "Invalid unique id '{0}'".format(uid)
            raise raeting.YardError(emsg)
    '''
    pass


class PageError(RaetError):
    '''
       Exceptions in RAET page processing

       Usage:
            emsg = "Invalid page kind '{0}'".format(kind)
            raise raeting.PageError(emsg)
    '''
    pass
